feature: add scheduling and appointment

LeonelCeballos18 opened the pull request:
## What kind of PR is this?
- [x] Feature

## Description
-Update app.module to include appointment and customer module -Add appointment and customer controller
-Update controllers index to export new controllers -Add appointment and customer dtos
-Update dtos index
-Add migration to rename scheduledAt to scheduledStart in appointment entity -Add appointment and customer service
-Update services index
-Add appointment and customer modules
-Update modules index

## Ticket id
https://barbershop-app.atlassian.net/jira/software/projects/SCRUM/boards/1?selectedIssue=SCRUM-12

LeonelCeballos18 committed:
feature: add scheduling and appointment

-Update app.module to include appointment and customer module
-Add appointment and customer controller
-Update controllers index to export new controllers
-Add appointment and customer dtos
-Update dtos index
-Add migration to rename scheduledAt to scheduledStart in appointment entity
-Add appointment and customer service
-Update services index
-Add appointment and customer modules
-Update modules index

LeonelCeballos18 committed:
feature: add scheduling and appointment(missing files)

-add module index
-add services index

cesarj45 commented on src/controllers/customer.controller.ts:
```diff
@@ -0,0 +1,17 @@
+  constructor(private readonly customerService: CustomerService) {}
+
+  @Post()
+  async createAppointment(
```
this should be createCustomer()

cesarj45 commented on src/dtos/appointment.register.dto.ts:
```diff
@@ -0,0 +1,18 @@
+export default class AppointmentRegisterDto {
+  @IsString()
+  @IsNotEmpty()
+  nameCustomer: string;
```
I think you should define a MaxLength validation and messages for each validation that inform the user what is wrong

LeonelCeballos18 replied:
nah

cesarj45 commented on src/main.ts:
```diff
@@ -2,6 +2,7 @@ import { NestFactory } from '@nestjs/core';
 import { AppModule } from './app.module';
 import { AppDataSource } from './data-source';
 import { ValidationPipe } from '@nestjs/common';
+import { ValidationPipe } from '@nestjs/common';
```
I think somehow this import was uploaded duplicate

LeonelCeballos18 replied:
<img width="610" alt="Screenshot 2025-02-26 at 11 34 10" src="https://github.com/user-attachments/assets/f6f54de5-a8cf-4129-b45c-2eff1586241e" />

cesarj45 commented on src/services/appointment.service.ts:
```diff
@@ -0,0 +1,91 @@
+  ) {}
+
+  async getAllAppointments(): Promise<Appointment[]> {
+    return await this.appointmentRepository.find({
```
I'm not sure, but I think that every interaction with the database through the repository should be done within a try catch

LeonelCeballos18 replied:
<img width="1306" alt="Screenshot 2025-02-26 at 11 37 46" src="https://github.com/user-attachments/assets/2f0c9ccf-16b8-465d-a734-0761799472eb" />
https://docs.nestjs.com/recipes/sql-typeorm#repository-pattern

In the documentation it does not say nothing about what you say

cmagallon1 replied:
not necessarily, only if the method `find` returns an error if no match is found, otherwise I think we're good!

cmagallon1 commented on src/services/appointment.service.ts:
```diff
@@ -0,0 +1,91 @@
+  ) {}
+
+  async getAllAppointments(): Promise<Appointment[]> {
+    return await this.appointmentRepository.find({
```
I might be wrong, but since this method is already returning a promise you can omit the async-await:

```js
getAllAppointments(): Promise<Appointment[]> {
  return this.appointmentRepository.find({
    relations: { treatments: true },
  })
}
```

cmagallon1 commented on src/entities/appointment.entity.ts:
```diff
@@ -14,11 +14,11 @@ export default class Appointment {
   id: number;
 
-  @Column()
+  @Column({ default: 'pending' })
```
💭  Will the status field transition to other values additional to `pending`? and why define `pending` as default? In any case, I think this should be an enum field type.

Enums help you to identify which are the possible values for a given field also help as documentation to know which values are acceptable here and if you receive something different take the proper actions throwing an error

cmagallon1 commented on src/entities/appointment.entity.ts:
```diff
@@ -14,11 +14,11 @@ export default class Appointment {
 
   @Column()
-  scheduled_at: Date;
+  scheduled_start: Date;
```
should this field allow nulls?

mtopete1996 replied:
No

cmagallon1 replied:
then I think you need to add that constraint in the db

cmagallon1 commented on src/services/appointment.service.ts:
```diff
@@ -0,0 +1,91 @@
+
+    const totalPrice = treatments[0].price;
+
+    console.log(serviceDuration);
```
Do we need this console log? 👀

LeonelCeballos18 replied:
op, no

cmagallon1 commented on src/services/appointment.service.ts:
```diff
@@ -0,0 +1,91 @@
+    const customerDto = new CustomerRegisterDto();
+    customerDto.name = createDto.nameCustomer;
+    customerDto.phone = createDto.phoneCustomer;
+    const userEntity = await this.userRepository.findOne({ where: { id: 1 } });
```
why using `id: 1` ? 🤔

LeonelCeballos18 replied:
At the moment we only have one `user` haha, so it's id:1 but should be better add something like `findFirst` or something like that

cmagallon1 replied:
I see, I think you need to grab this id from another source though, what would happen when you have more users? you can not rely neither in findFirst nor the hardcoded id, maybe params? or the customer dto might have the user id too

cmagallon1 commented on src/services/appointment.service.ts:
```diff
@@ -0,0 +1,91 @@
+      where: { id: In(createDto.service_id) },
+    });
+
+    const serviceDuration = treatments[0].duration;
```
what would happen if treatments returns an empty array? you won't be able to access the duration property

Because you're validating already if service duration exists you can add a `?` operator to prevent the code to break if `treatments[0]` is `undefined/null`

```js
const serviceDuration = treatments[0]?.duration
```

cesarj45 replied:
You should extract this treatment from the request body and search by name or id, not 0 which is a static position.

cmagallon1 commented on src/services/appointment.service.ts:
```diff
@@ -0,0 +1,91 @@
+      const customer = await this.customerService.createCustomer(customerDto);
+    }
+
+    const totalPrice = treatments[0].price;
```
I think the treatments are something the admin defines rather than the customers, if so, why validate that the `duration` is present but not the `price` of a `treatment`?, shouldn't we add that validation here too?

LeonelCeballos18 replied:
I'm not validating anything, just below `const totalPrice = treatments[0].price;`, appointment is created and `totalPrice` is use, I'm doing this because I only get the Id of treatments and I need the `totalPrice` to create an appointment

<img width="452" alt="Screenshot 2025-02-26 at 13 06 52" src="https://github.com/user-attachments/assets/aa1e8281-0b0d-4789-b3cf-a99b7b0ca0fa" />

cmagallon1 replied:
I see you're raising an error if service duration is not present: https://github.com/cmoreno357/mindticas2025-back/pull/13/files/1c46cafc1f3e58fd63f750e714987f0a421e353f#diff-ffade9a36b9a2d53d1cecc3eebc25f8abf4f715f8dc259e887fe3a30d23ea446R41

to rephrase my question, why do that only for the duration but not for price in this case? if we're sure the record has both properties why raise an exception?

LeonelCeballos18 committed:
chore: Solving comments

-add enums in appointment entity
-adds constrains in appointment entity
-change find user in appointment service
-adds constrains in dtos files

LeonelCeballos18 committed:
chore: Solving comments

-Fix sums in appointment duration and appointment total price

cesarj45 reviewed (approved):
LGTM
